File: java/src/main/java/Client/Communication.java
package Client;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.MulticastSocket;
import java.net.NetworkInterface;
import java.net.SocketTimeoutException;
import java.util.Enumeration;
import java.util.ArrayList;

import Protobuf.BallReplacement;
import Protobuf.Commands;
import Protobuf.Environment;
import Protobuf.Packet;
import Protobuf.Replacement;
import Protobuf.Robot;
import Protobuf.RobotReplacement;

/**
 * This class handles communication between the client and the robot soccer
 * simulation environment.
 */
public class Communication {
    public SharedObject<Boolean> rx;
    public SharedObject<byte[]> message;
    public SharedObject<Integer> messageLength;
    public SharedObject<Environment> environment;
    public SharedObject<ArrayList<Protobuf.Command>> commands;
    public SharedObject<byte[]> packet;
    public SharedObject<Integer> packetLength;
    public SharedObject<Boolean> replace;
    public SharedObject<Double[]> robotReplacePos;
    public SharedObject<Double[]> ballReplacePos;

    /**
     * Initializes the Communication class with shared objects for communication
     * data.
     */
    Communication() {
        rx = new SharedObject<>(false);
        message = new SharedObject<>(new byte[2048]);
        messageLength = new SharedObject<>(0);
        environment = new SharedObject<>(null);
        commands = new SharedObject<>(null);
        packet = new SharedObject<>(new byte[2048]);
        packetLength = new SharedObject<>(0);
        replace = new SharedObject<>(false);
        robotReplacePos = new SharedObject<>(new Double[3]);
        ballReplacePos = new SharedObject<>(new Double[2]);
    }

    /**
     * Gets the status of the reception flag.
     *
     * @return True if data has been received, false otherwise.
     */
    public boolean getRx() {
        return rx.Get();
    }

    /**
     * Receives frames from the multicast socket, handles reconnection on socket
     * errors,
     * and updates the shared communication objects accordingly.
     */
    private void receiveFrame() {
        try {
            byte[] receiveData = new byte[2048];
            InetAddress multicastAddress = InetAddress.getByName("224.0.0.1");
            int port = 10002;

            MulticastSocket socket = new MulticastSocket(new InetSocketAddress(port));

            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
            while (interfaces.hasMoreElements()) {
                NetworkInterface networkInterface = interfaces.nextElement();
                if (!networkInterface.isLoopback() && networkInterface.isUp()) {
                    socket.joinGroup(new InetSocketAddress(multicastAddress, port), networkInterface);
                }
            }

            while (true) {
                DatagramPacket receivePacket = new DatagramPacket(receiveData, receiveData.length);

                long startTime = System.currentTimeMillis(); // Get the current time
                boolean connected = true; // Assume we are initially connected

                // Try to receive data for 500ms or until connection is lost
                while (System.currentTimeMillis() - startTime < 500) {
                    try {
                        socket.setSoTimeout(500); // Set the socket timeout
                        socket.receive(receivePacket);
                        rx.Set(true); // Set the flag to true if data is received
                        message.Set(receivePacket.getData());
                        messageLength.Set(receivePacket.getLength());
                        break; // Exit the loop if data is received
                    } catch (SocketTimeoutException ste) {
                        // Timeout occurred, no data received, continue waiting
                    } catch (java.io.IOException ioe) {
                        // Connection lost, need to reconnect
                        connected = false;
                        break;
                    }
                }

                if (!connected) {
                    // Connection lost, reset the flag and attempt to reconnect
                    rx.Set(true);

                    // Leave the multicast group and rejoin with the new socket
                    NetworkInterface networkInterface = socket.getNetworkInterface();
                    if (networkInterface != null) {
                        socket.leaveGroup(new InetSocketAddress(multicastAddress, port), networkInterface);
                    }
                    socket.close();
                    socket = new MulticastSocket();
                    socket.setSoTimeout(500); // Set the socket timeout

                    Enumeration<NetworkInterface> newInterfaces = NetworkInterface.getNetworkInterfaces();
                    while (newInterfaces.hasMoreElements()) {
                        NetworkInterface newNetworkInterface = newInterfaces.nextElement();
                        if (!newNetworkInterface.isLoopback() && newNetworkInterface.isUp()) {
                            socket.joinGroup(new InetSocketAddress(multicastAddress, port), newNetworkInterface);
                        }
                    }
                }
            }
        } catch (Exception e) {
            System.out.println(e);
        }
    }

    /**
     * Decodes received messages into an Environment object and updates the shared
     * environment object.
     */
    private void decodeMessage() {
        while (true) {
            try {
                byte[] message = this.message.Get();
                int length = messageLength.Get();
                if (length == 0)
                    continue;
                byte[] actualMessage = new byte[length];
                System.arraycopy(message, 0, actualMessage, 0, length);
                Environment env = Environment.parseFrom(actualMessage);
                if (env != null)
                    environment.Set(env);
                else {
                    System.out.println("env is null");
                }
            } catch (Exception ex) {
                // System.out.println(ex);
            }
        }
    }

    /**
     * Prints the received frame information periodically.
     * This method should be running in a separate thread to continuously print the
     * updated frame information.
     */
    private void printReceivedFrame() {
        while (true) {
            try {
                Thread.sleep(80);
            } catch (Exception e) {
            }
            Environment env = environment.Get();
            if (rx.Get() && env != null) {
                System.out.println("rx");
                Main.clearTerminal();
                System.out.println("Field:");
                System.out.println(
                        "Length: " + env.getField().getLength() +
                                "; Width: " + env.getField().getWidth());
                System.out.println(
                        "GoalDepth: " + env.getField().getGoalDepth() +
                                "; GoalWidth: " + env.getField().getGoalWidth());
                System.out.println("Ball:");
                double vx = Main.round(env.getFrame().getBall().getVx(), 4),
                        vy = Main.round(env.getFrame().getBall().getVx(), 4), vnorm = Math.sqrt(vx *
                                vx + vy * vy);
                System.out.println(
                        "x: " + Main.round(env.getFrame().getBall().getX(), 4) +
                                "; y: " + Main.round(env.getFrame().getBall().getY(), 4) +
                                "; vX: " + vx +
                                "; vY: " + vy +
                                "; |v|: " + vnorm);
                System.out.println("Blue Robots:");
                for (Protobuf.Robot robot : env.getFrame().getRobotsBlueList()) {
                    System.out.println(
                            "id: " + robot.getRobotId() +
                                    "; x: " + Main.round(robot.getX(), 4) +
                                    "; y: " + Main.round(robot.getY(), 4) +
                                    "; vx: " + Main.round(robot.getVx(), 4) +
                                    "; vy: " + Main.round(robot.getVy(), 4));
                }
                System.out.println("Yellow Robots:");
                for (Protobuf.Robot robot : env.getFrame().getRobotsYellowList()) {
                    System.out.println(
                            "id: " + robot.getRobotId() +
                                    "; x: " + Main.round(robot.getX(), 4) +
                                    "; y: " + Main.round(robot.getY(), 4) +
                                    "; vx: " + Main.round(robot.getVx(), 4) +
                                    "; vy: " + Main.round(robot.getVy(), 4));
                }
            }
        }
    }

    /**
     * Sends packages via UDP to a specified IP address and port.
     * This method sends the data stored in the shared packet object.
     */
    private void sendPackage() {
        String ipAddress = Main.parameters.get("address").asText();
        int port = 20011;
        while (true) {
            try {
                Thread.sleep(14);
            } catch (Exception ex) {
            }
            byte[] packetArray = this.packet.Get();
            if (packetArray.length == 0)
                continue;
            // for (int i = 0; i < packetArray.length; i++) {
            // System.out.print(String.format("%02x", packetArray[i]));
            // }
            // System.out.print('\n');
            try (DatagramSocket socket = new DatagramSocket()) {
                InetAddress address = InetAddress.getByName(ipAddress);
                DatagramPacket packet = new DatagramPacket(packetArray, packetArray.length, address, port);
                socket.send(packet);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * Creates a default replacement message for the environment.
     * This message is used when replacements are needed.
     *
     * @return A Replacement message with default robot and ball positions.
     */
    private Replacement DefaultReplacement() {
        return Replacement.newBuilder()
                .setBall(BallReplacement.newBuilder().setX(ballReplacePos.Get()[0]).setY(ballReplacePos.Get()[1])
                        .build())
                .addRobots(RobotReplacement.newBuilder()
                        .setPosition(Robot.newBuilder().setX(robotReplacePos.Get()[0]).setY(robotReplacePos.Get()[1])
                                .setRobotId(0).setOrientation(robotReplacePos.Get()[2])
                                .build())
                        .setYellowteam(false).setTurnon(true).build())
                .addRobots(RobotReplacement.newBuilder()
                        .setPosition(Robot.newBuilder().setX(-0.6).setY(0.9).setRobotId(1).setOrientation(-90)
                                .build())
                        .setYellowteam(false).setTurnon(false).build())
                .addRobots(RobotReplacement.newBuilder()
                        .setPosition(Robot.newBuilder().setX(-0.3).setY(0.9).setRobotId(2).setOrientation(-90)
                                .build())
                        .setYellowteam(false).setTurnon(false).build())
                .addRobots(RobotReplacement.newBuilder()
                        .setPosition(
                                Robot.newBuilder().setX(1.5).setY(0).setRobotId(0).setOrientation(180).build())
                        .setYellowteam(true).setTurnon(true).build())
                .addRobots(RobotReplacement.newBuilder()
                        .setPosition(Robot.newBuilder().setX(0.3).setY(0.9).setRobotId(1).setOrientation(-90)
                                .build())
                        .setYellowteam(true).setTurnon(false).build())
                .addRobots(RobotReplacement.newBuilder()
                        .setPosition(Robot.newBuilder().setX(0.6).setY(0.9).setRobotId(2).setOrientation(-90)
                                .build())
                        .setYellowteam(true).setTurnon(false).build())
                .build();
    }

    /**
     * Encodes commands into a Packet and sends it.
     * This method constructs a Packet message containing commands and,
     * if needed, a replacement message, and sends it via UDP.
     */
    private void encodeMessage() {
        while (true) {
            try {
                Commands.Builder commandsBuilder = Commands.newBuilder();
                ArrayList<Protobuf.Command> commandsArray = this.commands.Get();
                if (commandsArray == null)
                    continue;
                // System.out.println(commandsArray.get(0).getWheelLeft());
                for (Protobuf.Command command : commandsArray) {
                    commandsBuilder.addRobotCommands(command);
                }
                Commands commands = commandsBuilder.build();
                Packet.Builder packetBuilder = Packet.newBuilder();
                packetBuilder = packetBuilder.setCmd(commands);
                if (replace.Get()) {
                    // System.out.println("replace");
                    packetBuilder = packetBuilder.setReplace(DefaultReplacement());
                }
                Packet packet = packetBuilder.build();
                byte[] serialized = packet.toByteArray();
                this.packet.Set(serialized);
            } catch (Exception ex) {
                System.out.println(ex);
            }
        }
    }

    /**
     * Starts the receiving threads for handling communication.
     * This method creates and starts threads for receiving, decoding, and printing
     * frames.
     */
    public void startReceiving() {
        Thread[] threads = {
                new Thread(this::receiveFrame),
                new Thread(this::decodeMessage),
                // new Thread(this::printReceivedFrame),
        };
        for (Thread thread : threads)
            thread.start();

        try {
            Thread.sleep(Long.MAX_VALUE);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        for (Thread thread : threads)
            thread.interrupt();
        try {
            for (Thread thread : threads)
                thread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    /**
     * Starts the sending threads for handling communication.
     * This method creates and starts threads for sending and encoding messages.
     */
    public void startSending() {
        Thread senderThread = new Thread(this::sendPackage);
        Thread encoderThread = new Thread(this::encodeMessage);

        senderThread.start();
        encoderThread.start();

        try {
            Thread.sleep(Long.MAX_VALUE);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        // Interrupt and join the threads
        senderThread.interrupt();
        encoderThread.interrupt();
        try {
            senderThread.join();
            encoderThread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

}

File: python/main/communication.py
import traceback
import threading
import os
import socket
from FiraSim_message_pb2 import Environment, Command, BallReplacement, RobotReplacement, Robot, Replacement, Commands, Packet
from shared_object import SharedObject
import time
import json

class Communication:
    def __init__(self):
        self.rx = SharedObject(False)
        self.message = SharedObject(bytearray(2048))
        self.message_length = SharedObject(0)
        self.environment = SharedObject(None)
        self.commands = SharedObject(None)
        self.packet = SharedObject(bytearray(2048))
        self.packet_length = SharedObject(0)
        self.replace = SharedObject(False)
        self.robot_replace_pos = SharedObject([0.0, 0.0, 0.0])
        self.ball_replace_pos = SharedObject([0.0, 0.0])
    def receive_frame(self):
        multicast_group = '224.0.0.1'
        port = 10002

        receive_data = bytearray(2048)
        multicast_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        multicast_socket.bind((multicast_group, port))

        multicast_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        multicast_socket.settimeout(0.5)  # Set the socket timeout
        try:
            while True:
                start_time = time.time()  # Get the current time
                connected = True  # Assume we are initially connected
                try:
                    #print('try get data')
                    data, _ = multicast_socket.recvfrom(2048)
                    # Set the flag to True if data is received
                    self.rx.set_value(True)
                    self.message.set_value(data)
                    self.message_length.set_value(len(data))
                except socket.timeout:
                    print('timeout')
                except socket.error as e:
                    # Connection lost, need to reconnect
                    print("Socket error:", e)
                    connected = False

                if not connected:
                    # Connection lost, reset the flag and attempt to reconnect
                    self.rx.set_value(True)
                    multicast_socket.close()
                    multicast_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
                    multicast_socket.bind(('', port))
                    multicast_socket.settimeout(0.5)  # Set the socket timeout
                    multicast_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        except KeyboardInterrupt:
            multicast_socket.close()

    def decode_message(self):
        while True:
            try:
                message = self.message.get_value()
                length = self.message_length.get_value()
                if length == 0:
                    continue

                actual_message = message[:length]
                env = Environment()

                try:
                    env.ParseFromString(actual_message)
                    if env:
                        self.environment.set_value(env)
                    else:
                        print("Environment is null")
                except Exception as ex:
                    pass
            except Exception as ex:
                print(f"Error decoding the message: {ex}")
                traceback.print_exc()

    def print_received_frame(self):
        while True:
            try:
                time.sleep(0.08)  # Sleeping for 80 milliseconds
                os.system('clear')
                env = self.environment.get_value()  # Assuming self.environment is a SharedObject
                if self.rx.get_value() and env is not None:
                    print("rx")
                    print("Field:")
                    print(f"Length: {env.field.length}")
                    print(f"Width: {env.field.width}")
                    print(f"GoalDepth: {env.field.goal_depth}")
                    print(f"GoalWidth: {env.field.goal_width}")

                    print("Ball:")
                    ball = env.frame.ball
                    vx = round(ball.vx, 4)
                    vy = round(ball.vy, 4)
                    vnorm = (vx ** 2 + vy ** 2) ** 0.5
                    print(f"x: {round(ball.x, 4)}; y: {round(ball.y, 4)}; vX: {vx}; vY: {vy}; |v|: {vnorm}")

                    print("Blue Robots:")
                    for robot in env.frame.robots_blue:
                        print(f"id: {robot.robot_id}; x: {round(robot.x, 4)}; y: {round(robot.y, 4)}; vx: {round(robot.vx, 4)}; vy: {round(robot.vy, 4)}")

                    print("Yellow Robots:")
                    for robot in env.frame.robots_yellow:
                        print(f"id: {robot.robot_id}; x: {round(robot.x, 4)}; y: {round(robot.y, 4)}; vx: {round(robot.vx, 4)}; vy: {round(robot.vy, 4)}")
            except Exception as e:
                pass
    def send_package(self):
        parameters = json.loads(open("../resources/parameters.json").read())
        ip_address = parameters["address"]
        port = 20011
        while True:
            try:
                time.sleep(0.014)  # 14 milliseconds
            except Exception as ex:
                pass
            packet_array = self.packet.get_value()
            if len(packet_array) == 0:
                continue
            # Uncomment the following lines to print the packet data in hex format
            # for i in packet_array:
            #     print(format(i, '02x'), end='')
            # print()

            with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
                address = (ip_address, port)
                sock.sendto(bytes(packet_array), address)
    
    def default_replacement(self):
        ball_replacement = BallReplacement(
            x=self.ball_replace_pos.get_value()[0],
            y=self.ball_replace_pos.get_value()[1],
            vx=0,
            vy=0
        )

        robots_replacement = [
            RobotReplacement(
                position=Robot(
                    robot_id=0,
                    x=self.robot_replace_pos.get_value()[0],
                    y=self.robot_replace_pos.get_value()[1],
                    orientation=self.robot_replace_pos.get_value()[2],
                    vx=0,
                    vy=0,
                    vorientation=0
                ),
                yellowteam=False,
                turnon=True
            ),
            RobotReplacement(
                position=Robot(
                    robot_id=1,
                    x=-0.6,
                    y=0.9,
                    orientation=-90,
                    vx=0,
                    vy=0,
                    vorientation=0
                ),
                yellowteam=False,
                turnon=False
            ),
            RobotReplacement(
                position=Robot(
                    robot_id=2,
                    x=-0.3,
                    y=0.9,
                    orientation=-90,
                    vx=0,
                    vy=0,
                    vorientation=0
                ),
                yellowteam=False,
                turnon=False
            ),
            RobotReplacement(
                position=Robot(
                    robot_id=0,
                    x=1.5,
                    y=0,
                    orientation=180,
                    vx=0,
                    vy=0,
                    vorientation=0
                ),
                yellowteam=True,
                turnon=True
            ),
            RobotReplacement(
                position=Robot(
                    robot_id=1,
                    x=0.3,
                    y=0.9,
                    orientation=-90,
                    vx=0,
                    vy=0,
                    vorientation=0
                ),
                yellowteam=True,
                turnon=False
            ),
            RobotReplacement(
                position=Robot(
                    robot_id=2,
                    x=0.6,
                    y=0.9,
                    orientation=-90,
                    vx=0,
                    vy=0,
                    vorientation=0
                ),
                yellowteam=True,
                turnon=False
            )
        ]

        replacement = Replacement(
            ball=ball_replacement,
            robots=robots_replacement
        )
        return replacement

    def encode_message(self):
        while True:
            try:
                commands_array = self.commands.get_value()
                if commands_array is None:
                    continue

                # Process Command messages
                robot_commands = []
                for command in commands_array:

                    robot_command = Command(
                        id=command.id,
                        yellowteam=command.yellowteam,
                        wheel_left=command.wheel_left,
                        wheel_right=command.wheel_right
                    )
                    robot_commands.append(robot_command)

                commands = Commands(robot_commands=robot_commands)

                packet = Packet(cmd=commands)
                
                if self.replace.get_value():
                    replacement = self.default_replacement()
                    packet.replace.CopyFrom(replacement)

                # Serialize the Packet
                serialized_packet = packet.SerializeToString()
                self.packet.set_value(serialized_packet)

            except Exception as ex:
                traceback.print_exc()
    def start_receiving(self):
        threads = [
            threading.Thread(target=self.receive_frame),
            threading.Thread(target=self.decode_message),
           # threading.Thread(target=self.print_received_frame)
        ]
        for thread in threads:
            thread.start()
        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            for thread in threads:
                thread.join()

    def start_sending(self):
        sender_thread = threading.Thread(target=self.send_package)
        encoder_thread = threading.Thread(target=self.encode_message)

        sender_thread.start()
        encoder_thread.start()

        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            sender_thread.join()
            encoder_thread.join()

# Usage
#comm = Communication()
#comm.startReceiving()
#comm.startSending()
